### ISSUE TITLE
Using 'Search selected in PyQGIS docs' removes selected text in Python Console(s)

### ISSUE THREAD

**Issue by rduivenvoorde** [qgis/QGIS#38942]
In current master if I use the  'Search selected in PyQGIS docs', from the editors context menu, on a word in the python console or editor, the selected text is DELETED from the line.

I will have a look into this myself too, but as said, I do not think I introduced it myself, as I only use:

``` self.selectedText() ```
see:
https://github.com/qgis/QGIS/blob/master/python/console/console_base.py#L159

which according to docs should just return a string, not remove it is it??

https://docs.huihoo.com/pyqt/QScintilla2/classQsciScintilla.html#ac3a47e883108c0bb8fecaa76e91d2267

To be honest, I earlier often had the feeling that if I thought to 'copy' something in the editor, that it actually 'cut' something (so the line or word was removed). First I thought I hit the wrong buttons (I use ctrl-V often, so thought to accidently hit ctrl-X), but after some time it is pretty obvious that there seems (to me) something wrong with either the key bindings OR the copy/selection functionality in the QsciScintilla panels ???

**Comment by rduivenvoorde** [comment]
Some more info: I cannot reproduce this on my Windows machine...

And IF I see this in the python console, and I have python editors open, and select some text in one of them (to copy it), only hitting the ctrl-button (so not yet ctrl-C) removes the whole selection there too....

More version info:

QGIS version | 3.15.0-Master | QGIS code revision | a15fff158a
-- | -- | -- | --
Compiled against Qt | 5.14.2 | Running against Qt | 5.14.2
Compiled against GDAL/OGR | 3.1.3 | Running against GDAL/OGR | 3.1.3
Compiled against GEOS | 3.8.1-CAPI-1.13.3 | Running against GEOS | 3.8.1-CAPI-1.13.3
Compiled against SQLite | 3.33.0 | Running against SQLite | 3.33.0
PostgreSQL Client Version | 12.4 (Debian 12.4-1) | SpatiaLite Version | 5.0.0
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.2
Compiled against PROJ | 7.1.1 | Running against PROJ | Rel. 7.1.1, September 1st, 2020
OS Version | Debian GNU/Linux bullseye/sid | This copy of QGIS writes debugging output.
Active python plugins | timemanager; nominatim_locator_filter; pdokservicesplugin; plugin_reloader; FeatureGridCreator; quick_map_services; qgis-maptiler-plugin

**Comment by nyalldawson** [comment]
I was also seeing this previously, but can't anymore. I think I'm missing some pre-requisite step which triggers this bug. Can you clarify a workflow from QGIS startup which triggers it?

**Comment by rduivenvoorde** [comment]
WIth me I can just create a fresh profiles path (to be sure there are no plugins or scripts interferring):

```
qgis --profiles-path /tmp/aaa
```

Then without creating a project or layer, (so still seeing News), I open de Python Console, and then type:

```
iface.mapCanvas()
```
there.
Select 'mapCanvas', and the use the context menu item. Gone is the word 'mapCanvas'...

**Comment by rduivenvoorde** [comment]
ping @slarosa (as you are deep into the console code :-) can you reproduce this? 
Or do you have a clue where to look?
I hesitate to keep it in in this state, although Windows (our biggest user base probably) do not have the issue...

**Comment by slarosa** [comment]
> ping @slarosa (as you are deep into the console code :-) can you reproduce this?
> Or do you have a clue where to look?
> I hesitate to keep it in in this state, although Windows (our biggest user base probably) do not have the issue...

@rduivenvoorde very odd, I cannot reproduce the issue on my machine (MacOS) following the steps in https://github.com/qgis/QGIS/issues/38942#issuecomment-697235924 and I don't remember having ever saw such an issue :-(

reading previous comments, is it only occurring on Linux??

**Comment by rduivenvoorde** [comment]
@slarosa yes, I think so, both Nyall and me are on Linux (even a whole fresh envronment/profile). The nightly osgeo4w builds on Windows do not have this issue. I do not have master builds on other linux machines/distro's...

**Comment by slarosa** [comment]
> @slarosa yes, I think so, both Nyall and me are on Linux (even a whole fresh envronment/profile). The nightly osgeo4w builds on Windows do not have this issue. I do not have master builds on other linux machines/distro's...

The only interaction with the text that I can see is at line  https://github.com/qgis/QGIS/blob/55ee2935a21bb1bddbeaee51f0186c40ad943ced/python/console/console_base.py#L160

@rduivenvoorde  Could you comment out the line and test it only inside the shell or editor pane?

(if it works then it is black magic :-))))

**Comment by rduivenvoorde** [comment]
@slarosa found that earlier too, and tried it. Also copied the string by adding something in front of it... 
But still deleting the selection.

**Comment by rduivenvoorde** [comment]
@nyalldawson Just found out that I have the same issue when I copy text (with ctrl-c) from the expression dialog:

![Screenshot-20201003161527-874x540](https://user-images.githubusercontent.com/731673/94993793-af0b0b00-0593-11eb-8ab2-32d85ec1f7b1.png)

not always, but if I copy and then click in another window, the expression text is actually removed/cut.

**Comment by slarosa** [comment]
@rduivenvoorde Could execute the following code in console and then doing the same steps from the widget displayed?
```py
from qgis.PyQt.Qsci import QsciScintilla


class Shell(QsciScintilla):        
    
    def contextMenuEvent(self, e):
        menu = QMenu(self)
        action = menu.addAction("Open QGIS Site", self.qgisSite)
        action.setEnabled(False)
        if self.hasSelectedText():
            action.setEnabled(True)
        menu.exec_(self.mapToGlobal(e.pos()))
        
    def qgisSite(self):
        if self.hasSelectedText():
            QDesktopServices.openUrl(QUrl('https://qgis.org/'))

w = Shell()
w.show()
```
and if this works then replacing QsciScintilla instance with QgsCodeEditor in Shell class?

Maybe a qscintilla bug? I am using 2.11.4 version.

**Comment by rduivenvoorde** [comment]
@slarosa yep, same issue, even if I make the code even more QGIS agostic like:

```python
# same behaviour as with straight PyQt5
#from qgis.PyQt.Qsci import QsciScintilla
#rom qgis.PyQt.QtWidgets import QApplication, QMenu
#from qgis.PyQt.QtGui import QDesktopServices
#from qgis.PyQt.QtCore import QUrl

from PyQt5.Qsci import QsciScintilla
from PyQt5.QtWidgets import QApplication, QMenu
from PyQt5.QtGui import QDesktopServices
from PyQt5.QtCore import QUrl

import sys

class Shell(QsciScintilla):

    def contextMenuEvent(self, e):
        menu = QMenu(self)
        action = menu.addAction("Open QGIS Site", self.qgisSite)
        action.setEnabled(False)
        if self.hasSelectedText():
            action.setEnabled(True)
        menu.exec_(self.mapToGlobal(e.pos()))

    def qgisSite(self):
        if self.hasSelectedText():
            QDesktopServices.openUrl(QUrl('https://qgis.org/'))

app = QApplication(sys.argv)
w = Shell()
w.show()
sys.exit(app.exec_())
```

QGIS about box shows: 

QScintilla2 Version | 2.11.5
-- | --
Compiled against Qt | 5.14.2

I tried to install just PyQt 5.14.2 and your version QScintilla 2.11.4 ( now I see the when I issues this, I had 2.11.2 ):
```
python3 -m venv venv
source venv/bin/activate
pip install PyQt5==5.14.2
pip install QScintilla==2.11.4
```
but it showed all the same issue...

Maybe a Qt / Gnome thingie?

**Comment by nyalldawson** [comment]
A weird way I can always reproduce this:

1. Open python console, open script editor, open a script
2. Open pycharm
3. Focus pycharm window
4. Select text in qgis script editor
5. Press "ctrl" -- text is removed instantly!!

**Comment by Pedro-Murteira** [comment]
@rduivenvoorde hello, I'm not being able to reproduce this issue on QGIS 3.22.4 and 3.24.0. Did you come across this issue more recently?

**Comment by rduivenvoorde** [comment]
It's definitly still there, I hit it almost every day I work with QGIS

**Comment by YoannQDQ** [comment]
Resolved by via #52493
